src/index.js

```diff
@@ -40,33 +57,65 @@ module.exports = class Client {
     this[PRIVATE.OFFSETS] = new Map()
     this[PRIVATE.LOGGER] = createLogger({ level: logLevel, logCreator })
     this[PRIVATE.CLUSTER_RETRY] = retry
-    this[PRIVATE.CREATE_CLUSTER] = ({
-      metadataMaxAge,
+    this[PRIVATE.CREATE_BROKERPOOL] = ({
+      metadataMaxAge = DEFAULT_METADATA_MAX_AGE,
       allowAutoTopicCreation = true,
       maxInFlightRequests = null,
       instrumentationEmitter = null,
-      isolationLevel,
-    }) =>
-      new Cluster({
+    }) => {
+      const connectionBuilder = createConnectionBuilder({
         logger: this[PRIVATE.LOGGER],
         retry: this[PRIVATE.CLUSTER_RETRY],
-        offsets: this[PRIVATE.OFFSETS],
+        instrumentationEmitter,
         socketFactory,
         brokers,
         ssl,
         sasl,
         clientId,
         connectionTimeout,
-        authenticationTimeout,
-        reauthenticationThreshold,
         requestTimeout,
         enforceRequestTimeout,
-        metadataMaxAge,
-        instrumentationEmitter,
-        allowAutoTopicCreation,
         maxInFlightRequests,
+      })
+      return new BrokerPool({
+        logger: this[PRIVATE.LOGGER],
+        retry: this[PRIVATE.CLUSTER_RETRY],
+        connectionBuilder,
+        allowAutoTopicCreation,
+        authenticationTimeout,
+        reauthenticationThreshold,
+        metadataMaxAge,
+      })
+    }
+    this[PRIVATE.CREATE_CLUSTER] = ({
+      isolationLevel,
+      brokerPool,
+      instrumentationEmitter,
+      ...brokerPoolOptions
+    }) => {
+      if (
+        Object.entries(brokerPoolOptions).filter(([, value]) => typeof value !== 'undefined')
+          .length > 0 &&
+        brokerPool
+      ) {
+        // XXX: We could compare against the actual options of the provided pool ...
+        throw new Error('Incompatible options: brokerPool and broker pool creation options')
+      } else if (!brokerPool) {
+        brokerPool = this[PRIVATE.CREATE_BROKERPOOL]({
+          ...brokerPoolOptions,
+          instrumentationEmitter,
+        })
+      } else if (instrumentationEmitter) {
+        brokerPool.forwardInstrumentationEvents(instrumentationEmitter)
```

Comment:
What's up with this forwarding of the instrumentation events?

Reply:
I haven't fully understood the idea here, but just one thing that we need to be mindful of is that we aren't duplicating events now. For example, let's say we are sending metrics on network requests, and we have both a consumer and a producer instrumented that share a brokerpool, would they both receive an event (and would they receive "each other's" events?

Reply:
The problem that this is supposed to solve is that consumers/producers emit instrumentation events from the underlying connections, with slightly adjusted names. I want to stay compatible with the existing API whereever possible, so that the sharing really can be completely ignored unless you need to look into it.

> For example, let's say we are sending metrics on network requests, and we have both a consumer and a producer instrumented that share a brokerpool, would they both receive an event (and would they receive "each other's" events?

Good one! Right now they would receive the events for their own requests and also other requests ... definitely needs to be touched further to prevent that, probably by adding some reasonably smart filter somewhere.

Reply:
I looked a lot here whether it would be somehow possible to know which event really is "caused by", and therefore should get reported to a particular consumer/producer.

I couldn't find a good way to do it in general, but looking at the problem a bit:
1. Most "interesting" events are coming from the cluster and the consumer/producer itself, and these aren't shared.
2. The only events not coming from there are the ones from the request queue, and at that point we have really no link to the originating consumer/producer. But also: Events such as "got a response to the metadata request" would be interesting to all consumers/producers using the same pool, because they share that metadata. Trying to pull this apart to find only those then that have interest in the specific topic (or request) seems like a huge undertaking.

I think that when you share a connection pool you should be somewhat aware of the fact that network traffic goes through that pool coming from all clients that you share the pool with, WDYT?

Reply:
My only worry is that we would ruin metrics collection by emitting the same event twice. For example, let's say people are collecting metrics on request queue depth or request duration. If those events are dispatched to multiple instances, that could skew the metrics.

I might be wrong about this, as I don't know the details of what events are shared and which aren't, but it feels like something that's very easy to get wrong for our users. That said, I don't have a solution off the top of my head either. If the current structure of the event emitters makes it impossible to know which instance to propagate the events to, maybe we need to restructure it somehow.

I hope that I can properly review this PR this weekend, and maybe have some ideas. 🤞